CMakeLists.txt

```diff
@@ -25,6 +25,19 @@ else()
   set(LINK_GENEX TARGET_FILE_NAME)
 endif()
 
+function(apply_output_rules target)
+  set_property(TARGET ${target} PROPERTY "ARCHIVE_OUTPUT_DIRECTORY" "${CMAKE_CURRENT_BINARY_DIR}/lib")
+  set_property(TARGET ${target} PROPERTY "LIBRARY_OUTPUT_DIRECTORY" "${CMAKE_CURRENT_BINARY_DIR}/lib")
+  set_property(TARGET ${target} PROPERTY "RUNTIME_OUTPUT_DIRECTORY" "${CMAKE_CURRENT_BINARY_DIR}/bin")
+  foreach (build_config ${CMAKE_CONFIGURATION_TYPES})
+    string(TOUPPER ${build_config} CFG_TYPE)
+    set_property(TARGET ${target} PROPERTY "ARCHIVE_OUTPUT_DIRECTORY_${CFG_TYPE}" "${CMAKE_CURRENT_BINARY_DIR}/lib")
+    set_property(TARGET ${target} PROPERTY "LIBRARY_OUTPUT_DIRECTORY_${CFG_TYPE}" "${CMAKE_CURRENT_BINARY_DIR}/lib")
+    set_property(TARGET ${target} PROPERTY "RUNTIME_OUTPUT_DIRECTORY_${CFG_TYPE}" "${CMAKE_CURRENT_BINARY_DIR}/bin")
+  endforeach()
+  set_property(TARGET ${target} PROPERTY "BUILD_INCLUDE_DIRECTORY" "${CMAKE_CURRENT_BINARY_DIR}/include")
```

Comment:
@friendlyanon Is there a reasonable way to do this in a way that doesn't destroy multi-config generators?

What I was hoping to create with this change was to have it so that whether a developer invokes the jakt compiler via `jakt my_file.jakt` from the runtime directory or an installed location, they can expect it to find any built-in includes or runtime libraries.

Using both the change tagged here and the `copy_include_to_build.cmake` file generation below the goal is, for single config:
```
build/
   bin/
      jakt
      jakttest
      etc
   include/
      runtime/
         All the runtime headers needed to compile a jakt program
   lib/
     future location of std lib, runtime lib, etc
```
And then for each config of a multi-config generator, it would have 
```
build/
   Debug/
      bin/
      include/
      lib/
   Release/
      bin/
      include/
      lib/
```

I looked around and it seems like LLVM uses `${CMAKE_CURRENT_BINARY_DIR}/${CMAKE_CFG_INTDIR}/bin`
https://github.com/llvm/llvm-project/blob/a46154cb1cd09aa26bc803d8696e6e9283aac6a9/llvm/CMakeLists.txt#L383-L398

But doing that makes `Ninja Multi-Config` very angry. I assume it makes MsBuild less angry, though I haven't tried it yet on my Windows PC with VS 2022.

What tripped me up with just setting the [non-config-suffixed properies](https://cmake.org/cmake/help/latest/prop_tgt/RUNTIME_OUTPUT_DIRECTORY.html) is that now Ninja Multi-Config puts the files in
`build/bin/Debug` or, sometimes, `build/bin/Debug/bin` which is super not what I wanted. I want the application to find its resources via the equivalent of `$(dirname $(dirname <current exe path>))/include` by default, while keeping the switches there for people with weird setups.

Trying to use things like `$<CONFIG>` in the file copy for the include files also tripped me up. I couldn't figure out how to get it to create the structure above using Ninja Multi-Config, in a way that re-copies any runtime/ headers into a per-config include directory when they change. The fact that `$<CONFIG>` expands to a lower-cased name makes it super confusing how you're supposed to use it to extract these properties back out later...

Reply:
I would just do this:

```cmake
  set(prefix "")
  get_cmake_property(is_multi_config GENERATOR_IS_MULTI_CONFIG)
  if(is_multi_config)
    set(prefix "$<CONFIG>/")
  endif()

  set_property(TARGET "${target}" PROPERTY ARCHIVE_OUTPUT_DIRECTORY "${CMAKE_CURRENT_BINARY_DIR}/${prefix}lib")
  set_property(TARGET "${target}" PROPERTY LIBRARY_OUTPUT_DIRECTORY "${CMAKE_CURRENT_BINARY_DIR}/${prefix}lib")
  set_property(TARGET "${target}" PROPERTY RUNTIME_OUTPUT_DIRECTORY "${CMAKE_CURRENT_BINARY_DIR}/${prefix}bin")
```

---

> What tripped me up with just setting the [non-config-suffixed properies](https://cmake.org/cmake/help/latest/prop_tgt/RUNTIME_OUTPUT_DIRECTORY.html) is thatnow Ninja Multi-Config puts the files in `build/bin/Debug` or, sometimes, `build/bin/Debug/bin` which is super not what I wanted.

To quote the link:

> Multi-configuration generators ([Visual Studio](https://cmake.org/cmake/help/latest/manual/cmake-generators.7.html#visual-studio-generators), [Xcode](https://cmake.org/cmake/help/latest/generator/Xcode.html#generator:Xcode), [Ninja Multi-Config](https://cmake.org/cmake/help/latest/generator/Ninja%20Multi-Config.html#generator:Ninja%20Multi-Config)) append a per-configuration subdirectory to the specified directory unless a generator expression is used.

and `$<CONFIG>` indeed prevents that from happening.

---

I got nothing for the header situation. Maybe try [`cmake -E create_symlink`](https://cmake.org/cmake/help/latest/manual/cmake.1.html#:~:text=multiple%20input%20files.-,create_symlink,-%3Cold%3E%20%3Cnew%3E)?

Reply:
aha, I was indeed over-thinking the heck out of it. The hint for using create_symlink for the headers seems much easier than what I was trying to do before with copying whenever anything changes. Thanks!